p4-16/spec/P4-16-spec.mdk

```diff
@@ -7200,6 +7200,63 @@ The `@match` annotation, described in Section [#sec-select], is used
 to specify a `match_kind` value other than the default `match_kind` of
 `exact` for a field of a `value_set`.
 
+### Extern function/method annotations { #sec-extern-annotations }
+
+Various annotations may appear on extern function and method declarations
+to describe limitations on the behavior and interactions of those
+functions.  By default extern functions might have any effect on the
+environment of the P4 program and might interact in non-trivial ways.
+Since externs are architecture-specific and their behavior is known to
+the architecture definition, these annotations are not strictly necessary
+(an implementation can have knowledge of how externs interact based
+on their names built into it), but these annotations provide a uniform
+way of describing certain well-defined interactions (or their absence),
+allowing architecture-indepedent analysis of P4 programs.
+
+- `@pure` - Describes a function that depends solely on its `in` argument
+  values and has no effect other than its `out` argument values and return
+  value.  No hidden state is recorded between calls, and its value does
+  not depend on any hidden state that may be changed by other calls.
+  An example is the `hash` function which computes a deterministic
+  hash of its arguments.  A `@pure` function whose results are unused
+  may be safely eliminated with no adverse effects, and mulitple calls
+  with identical arguments may be combined into a single call.  `@pure`
+  functions may also be reoerdered with respect to other computations
+  that are not data dependent.
+
+- `@noSideEffects` - Weaker than `@pure` and describes a function that
+  does not change any hidden state, but may depend on hidden state.
+  Such a function may be dead code eliminated, and may be reordered
+  or combined with other `@noSideEffects` or `@pure` calls, but not
+  with other function calls that may have side effects that affect
+  the function.
+
+- `@localState` - When applied to an extern method, the extern method
+  implementation might read and/or write internal state, but all such
+  accessed state is limited to state associated directly with the extern
+  object instance on which the method is called. When applied to an extern
+  function, the accessible state is restricted to state associated directly
+  with this extern function. Method and function calls so annotated will
+  never read, nor write, global state of the architecture, nor the state
+  of any other extern object instance or function.
+
+- `@packetState` - Similar to `@localState`, except it is even more
```

Comment:
I wrote this comment on the compiler PR: 
 It looks to me like we may want this annotation to be on the extern itself rather than on each method.
 Does it ever make sense to have only some methods of an extern be `@packetState`?

Reply:
I do not know if it ever makes sense to have only some methods of an extern be `@packetState` and not others.

I can say that in v1model and PSA architectures, the only ones that exist where `@packetState` appears correct are:
+ some of the extern functions
+ for PSA's Checksum, Digest, and InternetChecksum externs, where it does apply to all methods of the extern, not only some of them.

Details are in this article, where you can look for yes's in the `@packetState` column of the two tables: https://github.com/jafingerhut/p4-guide/tree/master/extern-annotations

Again, before we try to make localState, localIndexedState, and packetState more precise in the spec, better to decide whether we should leave them out of the spec, at least for now, until and unless they are deemed useful for compiler optimization passes.

Reply:
I can't at the moment think of an extern where we might want different annotations on different methods, but I'm not sure moving it to the extern (requiring all methods of an extern to be the same) really buys us anything (and it makes the implementation a little more cumbersome.)